modules/tinymce/src/core/main/ts/api/html/SaxParser.ts

```diff
@@ -118,40 +118,67 @@ const isInvalidUri = (settings: SaxParserSettings, uri: string, tagName: string)
   }
 };
 
+interface LightweightTag {
+  readonly tagDepth: number;
+  readonly end: number;
+}
+
+// This finds a tag, but with very small amounts of detail - used only in findMatchingEndTagIndex
+// Instead of returning what type of tag it is, we just say "is it an open tag, a close tag, or a self-closing tag (like a br, or comment)"
+// And we also don't bother returning the start index of the tag, just the end
+const findLightweightTag = (html: string, startIndex: number, shortEndedElements: SchemaMap): LightweightTag | null => {
+  const startTagRegExp = /<([!?\/])?([A-Za-z0-9\-_:.]+)/;
+  const endTagRegExp = /^(?:\s(?:[^'">]+(?:"[^"]*"|'[^']*'))*[^"'>]*(?:"[^">]*|'[^'>]*)?|\s*|\/)>/;
+
+  while (true) {
+    const startMatch = startTagRegExp.exec(html.slice(startIndex));
+    if (startMatch === null) {
+      return null;
+    }
+    const endOfStart = startIndex + startMatch.index + startMatch[0].length;
+
+    if (startMatch[0] === '<!--') { // it's a comment
+      const length = findCommentEndIndex(html.slice(startIndex), false);
+      return { tagDepth: 0, end: endOfStart + length };
+    } else { // it's an element
+      const endMatch = endTagRegExp.exec(html.slice(startIndex));
+      if (Type.isNull(endMatch)) {
+        // We can skip through to the end of startMatch only because there's no way a "<" could appear halfway through "<name-of-tag"
+        startIndex += startMatch.index + startMatch[0].length;
+        continue;
+      }
+      const end = endOfStart + endMatch[0].length;
+
+      if (startMatch[1] === '/') {
+        return { tagDepth: -1, end };
+      } else if (Obj.has(shortEndedElements, startMatch[2])) {
+        return { tagDepth: 0, end };
+      } else {
+        return { tagDepth: 1, end };
+      }
+    }
+  }
+};
+
 /**
- * Returns the index of the end tag for a specific start tag. This can be
- * used to skip all children of a parent element from being processed.
+ * Returns the index of the matching end tag for a specific start tag. This can
+ * be used to skip all children of a parent element from being processed.
  *
  * @private
- * @method findEndTagIndex
+ * @method findMatchingEndTagIndex
  * @param {tinymce.html.Schema} schema Schema instance to use to match short ended elements.
  * @param {String} html HTML string to find the end tag in.
  * @param {Number} startIndex Index to start searching at should be after the start tag.
  * @return {Number} Index of the end tag.
  */
-const findEndTagIndex = (schema: Schema, html: string, startIndex: number): number => {
-  let count = 1, index, matches;
-
+const findMatchingEndTagIndex = (schema: Schema, html: string, startIndex: number): number => {
```

Comment:
Happy to revert this, but I thought it might be useful to stress that we're looking for matching tags here. The name it's exported with hasn't changed, and the old name wasn't equal to the exported name anyway.